File: arm_fCWT/fCWT.py
import numpy as np
import math

PI = 3.14159265358979323846264338327950288419716939937510582097494459072381640628620899862803482534211706798
sqrt2PI = 2.50662827463100050241576528
IPI4 = 0.75112554446


class Morlet:
    def __init__(self, bandwidth):
        self.four_wavelen = 0.9876
        self.fb = bandwidth
        self.fb2 = 2.0 * self.fb * self.fb
        self.ifb = 1.0 / self.fb
        self.imag_frequency = False
        self.doublesided = False
        self.mother = None

    def generate(self, size, scale=None, real=None, imag=None):
        if scale is None:
            # Frequency domain
            self.width = size
            toradians = (2 * PI) / size
            norm = np.sqrt(2 * PI) * IPI4
            self.mother = np.zeros(self.width)

            for w in range(self.width):
                tmp1 = 2.0 * (w * toradians) * self.fb - 2.0 * PI * self.fb
                tmp1 = -(tmp1 * tmp1) / 2
                self.mother[w] = norm * np.exp(tmp1)
        else:
            # Time domain
            self.width = self.getSupport(scale)
            norm = size * self.ifb / np.sqrt(2 * np.pi)

            for t in range(self.width * 2 + 1):
                tmp1 = (t - self.width) / scale
                tmp2 = np.exp(-(tmp1 * tmp1) / (self.fb2))

                real[t] = norm * tmp2 * np.cos(tmp1 * 2.0 * np.pi) / scale
                imag[t] = norm * tmp2 * np.sin(tmp1 * 2.0 * np.pi) / scale

    def getWavelet(self, scale, pwav, pn):
        w = self.getSupport(scale)
        real = np.zeros(max(w * 2 + 1, pn))
        imag = np.zeros(max(w * 2 + 1, pn))

        self.generate(pn, scale, real, imag)

        for t in range(pn):
            pwav[t] = complex(real[t], imag[t])

    def getSupport(self, scale):
        # Placeholder for getSupport implementation
        return int(scale * 10)  # Example implementation


class Scales:
    def __init__(self, wav, st, afs, af0, af1, afn):
        self.fs = afs
        self.scales = np.zeros(afn)
        self.fourwavl = wav.four_wavelen
        self.nscales = afn

        if st == "FCWT_LOGSCALES":
            self.calculate_logscale_array(2.0, wav.four_wavelen, afs, af0, af1, afn)
        elif st == "FCWT_LINSCALES":
            self.calculate_linscale_array(wav.four_wavelen, afs, af0, af1, afn)
        else:
            self.calculate_linfreq_array(wav.four_wavelen, afs, af0, af1, afn)

    def getScales(self, pfreqs, pnf):
        for i in range(pnf):
            pfreqs[i] = self.scales[i]

    def getFrequencies(self, pfreqs, pnf):
        for i in range(pnf):
            pfreqs[i] = self.fs / self.scales[i]

    def calculate_logscale_array(self, base, four_wavl, fs, f0, f1, fn):

        s0 = fs / f1
        s1 = fs / f0
        power0 = math.log(s0) / math.log(base)
        power1 = math.log(s1) / math.log(base)
        dpower = power1 - power0

        for i in range(fn):
            power = power0 + (dpower / (fn - 1)) * i
            self.scales[i] = base**power

    def calculate_linfreq_array(self, four_wavl, fs, f0, f1, fn):

        df = f1 - f0

        for i in range(fn):
            self.scales[fn - i - 1] = fs / (f0 + (df / fn) * i)

    def calculate_linscale_array(self, four_wavl, fs, f0, f1, fn):

        s0 = fs / f1
        s1 = fs / f0
        ds = s1 - s0

        for i in range(fn):
            self.scales[i] = s0 + (ds / fn) * i


if __name__ == "__main__":
    # Example usage of the Morlet and Scales classes
    bandwidth = 1  # Example bandwidth for the Morlet wavelet
    morlet_wavelet = Morlet(bandwidth)

    # Define scale type and frequency parameters
    scale_type = "FCWT_LINFREQS"
    afs = 1000  # Sampling frequency in Hz
    af0 = 0.1  # Start frequency in Hz
    af1 = 20  # End frequency in Hz
    afn = 10  # Number of scales

    # Initialize Scales object with the defined parameters
    scales = Scales(morlet_wavelet, scale_type, afs, af0, af1, afn)

    # Example: Get frequencies corresponding to the scales
    frequencies = np.zeros(afn)
    scales.getFrequencies(frequencies, afn)

    print("Frequencies corresponding to the scales:")
    print(frequencies)
    print("Scales:")
    print(scales.scales)
    scales.getScales(frequencies, afn)
    print("Scale:")
    print(frequencies)

File: arm_fCWT/fcwt.cc
//
//  fcwt.cpp
//  fCWT
//
//  Created by Lukas Arts on 21/12/2020.
//  Copyright © 2021 Lukas Arts.
/*Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

//
//  fcwt.cpp
//  fCWT-testing
//
//  Created by Lukas Arts on 21/12/2020.
//  Copyright © 2020 Lukas Arts.
/*Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "fcwt.h"

Morlet::Morlet(float bandwidth) {
    four_wavelen = 0.9876f;
    fb = bandwidth;
    fb2 = 2.0f * fb * fb;
    ifb = 1.0f / fb;
    imag_frequency = false;
    doublesided = false;
    mother = NULL;
}

void Morlet::generate(int size) {
    // Frequency domain, because we only need size. Default scale is always 2;
    width = size;

    float tmp1;
    float toradians = (2 * PI) / (float)size;
    float norm = sqrt(2 * PI) * IPI4;

    mother = (float *)malloc(sizeof(float) * width);

    // calculate array
    for (int w = 0; w < width; w++) {
        tmp1 = (2.0f * ((float)w * toradians) * fb - 2.0f * PI * fb);
        tmp1 = -(tmp1 * tmp1) / 2;
        mother[w] = (norm * exp(tmp1));
    }
}
void Morlet::generate(float *real, float *imag, int size, float scale) {
    // Time domain because we know size from scale
    float tmp1, tmp2;
    width = getSupport(scale);
    float norm = (float)size * ifb * IPI4;

    for (int t = 0; t < width * 2 + 1; t++) {
        tmp1 = (float)(t - width) / scale;
        tmp2 = exp(-(tmp1 * tmp1) / (fb2));

        real[t] = norm * tmp2 * cos(tmp1 * 2.0f * PI) / scale;
        imag[t] = norm * tmp2 * sin(tmp1 * 2.0f * PI) / scale;
    }
}

void Morlet::getWavelet(float scale, complex<float> *pwav, int pn) {
    int w = getSupport(scale);

    float *real = (float *)malloc(sizeof(float) * max(w * 2 + 1, pn));
    float *imag = (float *)malloc(sizeof(float) * max(w * 2 + 1, pn));
    for (int t = 0; t < max(w * 2 + 1, pn); t++) {
        real[t] = 0;
        imag[t] = 0;
    }

    generate(real, imag, pn, scale);

    for (int t = 0; t < pn; t++) {
        pwav[t].real(real[t]);
        pwav[t].imag(imag[t]);
    }

    delete real;
    delete imag;
};

//=============================================================//
//================== Scales ===================================//
//=============================================================//

Scales::Scales(Wavelet *wav, SCALETYPE st, int afs, float af0, float af1, int afn) {
    fs = afs;
    scales = (float *)malloc(afn * sizeof(float));
    fourwavl = wav->four_wavelen;
    nscales = afn;

    if (st == SCALETYPE::FCWT_LOGSCALES)
        calculate_logscale_array(2.0f, wav->four_wavelen, afs, af0, af1, afn);
    else if (st == SCALETYPE::FCWT_LINSCALES)
        calculate_linscale_array(wav->four_wavelen, afs, af0, af1, afn);
    else
        calculate_linfreq_array(wav->four_wavelen, afs, af0, af1, afn);
}

void Scales::getScales(float *pfreqs, int pnf) {
    for (int i = 0; i < pnf; i++) {
        pfreqs[i] = scales[i];
    };
};

void Scales::getFrequencies(float *pfreqs, int pnf) {
    for (int i = 0; i < pnf; i++) {
        pfreqs[i] = ((float)fs) / scales[i];
    };
};

void Scales::calculate_logscale_array(float base, float four_wavl, int fs, float f0, float f1, int fn) {
    // If a signal has fs=100hz and you want to measure [0.1-50]Hz, you need scales 2 to 1000;
    float nf0 = f0;
    float nf1 = f1;
    float s0 = (fs / nf1);
    float s1 = (fs / nf0);
    float power0 = log(s0) / log(base);
    float power1 = log(s1) / log(base);
    float dpower = power1 - power0;

    for (int i = 0; i < fn; i++) {
        float power = power0 + (dpower / (fn - 1)) * i;
        scales[i] = pow(base, power);
    }
}

void Scales::calculate_linfreq_array(float four_wavl, int fs, float f0, float f1, int fn) {
    float nf0 = f0;
    float nf1 = f1;
    // If a signal has fs=100hz and you want to measure [0.1-50]Hz, you need scales 2 to 1000;
    float df = nf1 - nf0;

    for (int i = 0; i < fn; i++) {
        scales[fn - i - 1] = (((float)fs) / (nf0 + (df / fn) * (float)i));
    }
}

void Scales::calculate_linscale_array(float four_wavl, int fs, float f0, float f1, int fn) {
    float nf0 = f0;
    float nf1 = f1;
    // If a signal has fs=100hz and you want to measure [0.1-50]Hz, you need scales 2 to 1000;
    float s0 = fs / nf1;
    float s1 = fs / nf0;
    float ds = s1 - s0;

    for (int i = 0; i < fn; i++) {
        scales[i] = (s0 + (ds / fn) * i);
    }
}

//=============================================================//
//================== FCWT =====================================//
//=============================================================//

void FCWT::daughter_wavelet_multiplication(float32_t *input, float32_t *output, const float32_t *mother, float scale, int isize, bool imaginary, bool doublesided) {
    float isizef = static_cast<float>(isize);
    float endpointf = fmin(isizef / 2.0, (isizef * 2.0) / scale);
    float step = scale / 2.0;
    int endpoint = static_cast<int>(endpointf);
    float maximum = isizef - 1;
    int s1 = isize - 1;

    for (int q1 = 0; q1 < endpoint; q1++) {
        float q = static_cast<float>(q1);
        float tmp = fmin(maximum, step * q);

        output[2 * q1] = input[2 * q1] * mother[static_cast<int>(tmp)];
        output[2 * q1 + 1] = input[2 * q1 + 1] * mother[static_cast<int>(tmp)] * (1 - 2 * imaginary);
    }

    if (doublesided) {
        for (int q1 = 0; q1 < endpoint; q1++) {
            float q = static_cast<float>(q1);
            float tmp = fmin(maximum, step * q);

            output[2 * (s1 - q1)] = input[2 * (s1 - q1)] * mother[static_cast<int>(tmp)] * (1 - 2 * imaginary);
            output[2 * (s1 - q1) + 1] = input[2 * (s1 - q1) + 1] * mother[static_cast<int>(tmp)];
        }
    }
}

void FCWT::fftbased(arm_cfft_instance_f32 &cfft_instance, float32_t *Ihat, float32_t *O1, float32_t *out, const float32_t *mother, int size, float scale, bool imaginary, bool doublesided) {
    // Generation of child wavelet function and multiplication with transformed input signal
    daughter_wavelet_multiplication(Ihat, O1, mother, scale, size, imaginary, doublesided);
    // Performing inverse FFT using ARM CMSIS-DSP
    arm_cfft_f32(&cfft_instance, O1, 1, 1);
    // Copying the result to the output buffer
    memcpy(out, O1, 2 * size * sizeof(float32_t));
}

void FCWT::cwt(float *pinput, int psize, complex<float> *poutput, Scales *scales) {
    arm_cfft_instance_f32 cfft_instance;
    size = psize;

    // Find the closest power of two
    const int nt = find2power(size);
    const int newsize = 1 << nt;

    // Initialize the intermediate result
    float32_t *Ihat = (float32_t *)pvPortMalloc(2 * newsize * sizeof(float32_t));
    float32_t *O1 = (float32_t *)pvPortMalloc(2 * newsize * sizeof(float32_t));
    memset(Ihat, 0, 2 * newsize * sizeof(float32_t));
    memset(O1, 0, 2 * newsize * sizeof(float32_t));

    // Initialize FFT plans
    status = arm_cfft_init_f32(&cfft_instance, newsize);
    // if (status != ARM_MATH_SUCCESS) {
    //     printf("CFFT initialization error!\n");
    // }

    // Calculating forward FFT
    float32_t *input_float32 = (float32_t *)pvPortMalloc(2 * newsize * sizeof(float32_t));
    // if (input_float32 == NULL) {
    //     printf("Memory allocation error!\n");
    // }
    memset(input_float32, 0, 2 * newsize * sizeof(float32_t));
    // Copy the input signal to the float32 array
    for (int i = 0; i < size; i++) {
        input_float32[2 * i] = pinput[i];  // The real part
        input_float32[2 * i + 1] = 0.0f;   // The imaginary part is zero
    }

    arm_cfft_f32(&cfft_instance, input_float32, 0, 1);
    memcpy(Ihat, input_float32, 2 * newsize * sizeof(float32_t));

    vPortFree(input_float32);

    // Generation of the mother wavelet function
    wavelet->generate(newsize);

    complex<float> *out = poutput;

    for (int i = 0; i < scales->nscales; i++) {
        // FFT based convolution in frequency domain
        fftbased(cfft_instance, Ihat, O1, (float32_t *)out, wavelet->mother, newsize, scales->scales[i], wavelet->imag_frequency, wavelet->doublesided);
        if (use_normalization) fft_normalize(out, newsize);
        out += size;
    }

    // Cleanup
    vPortFree(Ihat);
    vPortFree(O1);
}

void FCWT::fft_normalize(complex<float> *out, int size) {
    for (int i = 0; i < size; i++) {
        out[i] = out[i] / static_cast<float>(size);
    }
}